### PR TITLE
Remove dependency on sharetab file and refactor sharing logic

### PR DESCRIPTION
**Pull request by grwilson**
### Motivation and Context

The current implementation of 'sharenfs' and 'sharesmb' relies on
the use of the sharetab file. The use of this file is os-specific
and not required by linux or freebsd. Currently the code must
maintain updates to this file which adds complexity and presents
a significant performance impact when sharing many datasets. In
addition, concurrently running 'zfs sharenfs' command results in
missing entries in the sharetab file leading to unexpected failures.

### Description

This change removes the sharetab logic from the linux and freebsd
implementation of 'sharenfs' and 'sharesmb'. It still preserves an
os-specific library which contains the logic required for sharing
NFS or SMB. The following entry points exist in the vastly simplified
libshare library:

- sa_enable_share -- shares a dataset but may not commit the change
- sa_disable_share -- unshares a dataset but may not commit the change
- sa_is_shared -- determine if a dataset is shared
- sa_share_commit -- notify NFS/SMB subsystem to commit the shares
- sa_validate_shareopts -- determine if sharing options are valid

The sa_share_commit entry point is provided as a performance enhancement
and is not required. The sa_enable_share/sa_disable_share may commit
the share as part of the implementation. Libshare provides a framework
for both NFS and SMB but some operating systems may not fully support
these protocols or all features of the protocol.

NFS Operation:
For linux, libshare updates /etc/exports.d/zfs.exports to add
and remove shares and then commits the changes by invoking 'exportfs -r'.
This file, is automatically read by the kernel NFS implementation which
makes for better integration with the NFS systemd service. For FreeBSD,
libshare updates /etc/zfs/exports to add and remove shares and then
commits the changes by sending a SIGHUP to mountd.

SMB Operation:
For linux, libshare adds and removes files in /var/lib/samba/usershares
by calling the 'net' command directly. There is no need to commit the
changes. FreeBSD does not support SMB.

### Performance Results

To test sharing performance we created a pool with an increasing number
of datasets and invoked various zfs actions that would enable and disable
sharing. The performance testing was limited to NFS sharing. The following
tests were performed on an 8 vCPU system with 128GB and a pool comprised
of 4 50GB SSDs:

Scale testing:
- Share all filesystems in parallel -- zfs sharenfs=on <dataset> &
- Unshare all filesystems in parallel -- zfs sharenfs=off <dataset> &

Functional testing:
- share each filesystem serially -- zfs share -a
- unshare each filesystem serially -- zfs unshare -a
- reset sharenfs property and unshare -- zfs inherit -r sharenfs <pool>

For 'zfs sharenfs=on' scale testing we saw an average reduction in time
of 89.43% and for 'zfs sharenfs=off' we saw an average reduction in time
of 83.36%.

Functional testing also shows a huge improvement:
- zfs share -- 97.97% reduction in time
- zfs unshare -- 96.47% reduction in time
- zfs inhert -r sharenfs -- 99.01% reduction in time

The following charts show the results of scale testing. For each tests we create an increasing number of datasets and then invoke the respective 'sharenfs' command in the background.

baseline sharenfs=off and new sharenfs=off
basline sharenfs=on and new sharenfs=on

Signed-off-by: George Wilson <gwilson@delphix.com>
External-Issue: DLPX-68690

### CODE REVIEW

**Comment by grwilson** [comment]
I have updated the PR since one of the commits was missed and I also had to remove some leftover debugging code. I have also fixed up the lint errors and style errors that were reported.

**Comment by freqlabs** [comment]
I think the zfs_share tests are being skipped on FreeBSD in setup. There were some issues with my attempt at libtest is_shared hanging in some situations so I punted at the time. I will see if I can manually run the test to verify functionality for now.

**Comment by freqlabs** [comment]
I'm finding this in /etc/zfs/exports when running zfs_share tests after fixing libtest and the zfs_share setup to run on FreeBSD (before the concurrent shares test):
```
# !!! DO NOT EDIT THIS FILE MANUALLY !!!

S FILE MANUALLY !!!
S FILE MANUALLY !!!
S FILE MANUALLY !!!
 FILE MANUALLY !!!
NUALLY !!!
/var/tmp/testdir1       on
 FILE MANUALLY !!!
 FILE MANUALLY !!!
S FILE MANUALLY !!!
 FILE MANUALLY !!!
```
Needless to say, the tests do not go well :)

That is with this PR applied. I don't see this behavior on master.

File: lib/libshare/libshare.c [hunk 1]
```diff
@@ -78,477 +95,111 @@ register_fstype(const char *name, const sa_share_ops_t *ops)
...
-		}
-	}
-
-	free(impl_share->dataset);
-	impl_share->dataset = dataset_dup;
-
-	rc = SA_INVALID_PROTOCOL;
+		return (SA_NO_MEMORY);
 
+	(void) libshare_lock();
```
Comment (ahrens) [inline lib/libshare/libshare.c:117]: Is there a comment somewhere describing this locking?  I'm wondering why we need locking at this layer.  I can imagine the impl's update_shareopts() needing locking, but do we also need to ensure that update_shareopts() and enable_share() are both called before any other process calls `fstype->ops->...()`?  (I'm assuming the list of types is static so the locking is not for that.)

File: lib/libshare/libshare.c [hunk 2]
```diff
@@ -78,477 +95,111 @@ register_fstype(const char *name, const sa_share_ops_t *ops)
...
 	}
+	free_share(impl_share);
+	return (ret);
+}
 
-	update_sharetab(impl_share->handle);
-
-	return (found_protocol ? ret : SA_INVALID_PROTOCOL);
+void
+sa_share_commit(void)
```
Comment (ahrens) [inline lib/libshare/libshare.c:196]: Since sa_enable_share() is for a specific protocol, I think it would make sense for this to also act on a specific protocol.  Although maybe that would make the consumer's jobs more complicated, depending on how many layers of abstraction are between sa_enable_share() and sa_share_commit().

Comment (ahrens) [inline lib/libshare/libshare.c:196]: The other functions seem to follow the naming convention `sa_VERB_OBJECT()`, so how about:
```suggestion
sa_commit_shares(void)
```

File: lib/libshare/libshare_impl.h [hunk 3]
```diff
@@ -22,36 +22,36 @@
...
-	char *dataset;
+	char *sa_mountpoint;
+	char *sa_zfsname;
 
 	sa_share_fsinfo_t *fsinfo; /* per-fstype information */
 } *sa_share_impl_t;
 
 #define	FSINFO(impl_share, fstype) (&(impl_share->fsinfo[fstype->fsinfo_index]))
 
+#define	LIBSHARE_LOCK_FILE	"/etc/zfs/libshare.lck"
```
Comment (ahrens) [inline lib/libshare/libshare_impl.h:41]: is `.lck` a pre-existing convention?  If not, I think spelling out `.lock` would be helpful.

File: lib/libshare/os/linux/nfs.c [hunk 4]
```diff
@@ -23,34 +23,34 @@
...
 typedef int (*nfs_shareopt_callback_t)(const char *opt, const char *value,
     void *cookie);
 
-typedef int (*nfs_host_callback_t)(const char *sharepath, const char *host,
-    const char *security, const char *access, void *cookie);
+typedef int (*nfs_host_callback_t)(const char *sharepath, const char *filename,
+    const char *host, const char *security, const char *access, void *cookie);
 
 /*
  * Invokes the specified callback function for each Solaris share option
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:56]: Are these really `Solaris` share options?

File: lib/libshare/os/linux/nfs.c [hunk 5]
```diff
@@ -226,66 +233,57 @@ get_linux_hostspec(const char *solaris_hostspec, char **plinux_hostspec)
...
-
-		exit(1);
+		return (SA_NO_MEMORY);
 	}
 
-	sprintf(hostpath, "%s:%s", linuxhost, sharepath);
-
+	sprintf(buffer, fmt, sharepath, linuxhost,
+	    security, access_opts, linux_opts);
 	free(linuxhost);
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:267]: I think you can use `asprintf()` to simplify this:
```
char *buffer;
int err = asprintf(&buffer, "%s %s(sec=%s,%s,%s)\n", ...);
free(linuxhost);
if (err == -1)
  return (SA_NO_MEMORY);
```

File: lib/libshare/os/linux/nfs.c [hunk 6]
```diff
@@ -226,66 +233,57 @@ get_linux_hostspec(const char *solaris_hostspec, char **plinux_hostspec)
...
+	int fd = open(filename, O_RDWR | O_APPEND | O_CREAT, 0644);
+	if (fd == -1) {
+		fprintf(stderr, "failed to open %s file: %s", filename,
+		    strerror(errno));
+		free(buffer);
+		return (SA_SYSTEM_ERR);
+	}
 
-	if (rc < 0)
+	if (write(fd, buffer, strlen(buffer)) == -1) {
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:277]: Did you consider using STDIO routines (C standard buffered i/o) to (potentially) simplify this?
```
FILE *f = fopen(filename, "a+");
// handle error
fprintf(f, "%s %s(sec=%s,%s,%s)\n", ...);
// handle error
fclose(f);
// handle error
```

File: lib/libshare/os/linux/nfs.c [hunk 7]
```diff
@@ -226,66 +233,57 @@ get_linux_hostspec(const char *solaris_hostspec, char **plinux_hostspec)
...
+		perror("failed to write exports file");
+		close(fd);
+		free(buffer);
 		return (SA_SYSTEM_ERR);
-	else
-		return (SA_OK);
+	}
+	fsync(fd);
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:283]: Why do we need fo fsync() here, rather than in nfs_export_all()?  Seems like it would perform better to do it just once per logical operation.  If we crash in the middle, before nfs_export_all(), it seems like the file would be half-modified even with this fsync().

File: lib/libshare/os/linux/nfs.c [hunk 8]
```diff
@@ -634,114 +602,33 @@ nfs_update_shareopts(sa_share_impl_t impl_share, const char *resource,
...
 
 static const sa_share_ops_t nfs_shareops = {
 	.enable_share = nfs_enable_share,
 	.disable_share = nfs_disable_share,
+	.is_shared = nfs_is_share_active,
 
 	.validate_shareopts = nfs_validate_shareopts,
 	.update_shareopts = nfs_update_shareopts,
 	.clear_shareopts = nfs_clear_shareopts,
+	.commit_shares = nfs_export_all,
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:616]: rename the function to match the op name
```suggestion
	.commit_shares = nfs_commit_shares,
```

File: lib/libshare/os/linux/nfs.c [hunk 9]
```diff
@@ -634,114 +602,33 @@ nfs_update_shareopts(sa_share_impl_t impl_share, const char *resource,
...
 nfs_clear_shareopts(sa_share_impl_t impl_share)
 {
-	free(FSINFO(impl_share, nfs_fstype)->shareopts);
 	FSINFO(impl_share, nfs_fstype)->shareopts = NULL;
 }
 
 static const sa_share_ops_t nfs_shareops = {
 	.enable_share = nfs_enable_share,
 	.disable_share = nfs_disable_share,
+	.is_shared = nfs_is_share_active,
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:611]: rename the function to match the op name
```suggestion
	.is_shared = nfs_is_shared,
```

File: lib/libshare/os/linux/nfs.c [hunk 10]
```diff
@@ -407,77 +405,124 @@ get_linux_shareopts(const char *shareopts, char **plinux_opts)
...
 {
-	char *shareopts, *linux_opts;
-	int rc;
+	char *tmpfile = NULL;
 
-	if (!nfs_available()) {
+	if (asprintf(&tmpfile, "%s%s", ZFS_EXPORTS_FILE, ".XXXXXX") == -1) {
+		fprintf(stderr, "Unable to allocate temporary file\n");
+		return (NULL);
+	}
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:416]: Since you aren't using it here (which I think is fine), I think you can get rid of ZFS_EXPORTS_FILE_TEMP.

File: lib/libshare/os/linux/nfs.c [hunk 11]
```diff
@@ -407,77 +405,124 @@ get_linux_shareopts(const char *shareopts, char **plinux_opts)
...
+	/*
+	 * If the file doesn't exist then there is nothing more
+	 * we need to do.
+	 */
+	FILE *oldfp = fopen(ZFS_EXPORTS_FILE, "r");
+	if (oldfp == NULL)
 		return (SA_OK);
 
-	rc = get_linux_shareopts(shareopts, &linux_opts);
+	buf = (char *)malloc(buflen * sizeof (char));
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:456]: > `* sizeof (char)`

Aren't we standards compliant!

But I don't think the cast is necessary here.

File: lib/libshare/os/linux/nfs.c [hunk 12]
```diff
@@ -407,77 +405,124 @@ get_linux_shareopts(const char *shareopts, char **plinux_opts)
...
+				continue;
+			}
+		}
+		fprintf(newfp, "%s", buf);
+	}
 
-	free(linux_opts);
+	free(buf);
+
+	fflush(newfp);
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:477]: same question as for fsync - can't it be done in nfs_export_all() instead?

File: lib/libshare/os/linux/nfs.c [hunk 13]
```diff
@@ -407,77 +405,124 @@ get_linux_shareopts(const char *shareopts, char **plinux_opts)
...
+	fclose(newfp);
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:478]: I think we need to check for errors from fclose()

File: lib/libshare/os/linux/nfs.c [hunk 14]
```diff
@@ -407,77 +405,124 @@ get_linux_shareopts(const char *shareopts, char **plinux_opts)
...
+	return (SA_OK);
+}
+
+static int
+nfs_export_all(void)
+{
+	char *argv[3];
+	argv[0] = "/usr/sbin/exportfs";
+	argv[1] = "-r";
+	argv[2] = NULL;
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:490]:
```suggestion
char *argv[] = [
    "/usr/sbin/exportfs",
    "-r",
    NULL
];```

File: lib/libshare/os/linux/nfs.c [hunk 15]
```diff
@@ -407,77 +405,124 @@ get_linux_shareopts(const char *shareopts, char **plinux_opts)
...
 
+	int rc = libzfs_run_process(argv[0], argv, 0);
 	return (rc);
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:493]:
```suggestion
	return (libzfs_run_process(argv[0], argv, 0));
```

File: lib/libshare/os/linux/nfs.c [hunk 16]
```diff
@@ -407,77 +405,124 @@ get_linux_shareopts(const char *shareopts, char **plinux_opts)
...
-
-	if (rc < 0)
-		return (SA_SYSTEM_ERR);
-	else
-		return (SA_OK);
+	rc = foreach_nfs_host(impl_share, filename, nfs_add_entry, linux_opts);
+	free(linux_opts);
+	if (rc == 0) {
+		rc = nfs_fini_tmpfile(filename);
+	}
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:524]: if `rc` is nonzero, where do we `unlink()` and `free()` the `filename`?

File: lib/libshare/os/linux/nfs.c [hunk 17]
```diff
@@ -512,118 +554,44 @@ nfs_validate_shareopts(const char *shareopts)
...
 nfs_is_share_active(sa_share_impl_t impl_share)
 {
-	int fd;
-	char line[512];
-	char *tab, *cur;
-	FILE *nfs_exportfs_temp_fp;
-
-	if (!nfs_available())
+	size_t buflen = MAXPATHLEN;
+	char *buf = (char *)malloc(buflen * sizeof (char));
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:564]: same question about allocating the arg to getline()

File: lib/libshare/os/linux/nfs.c [hunk 18]
```diff
@@ -512,118 +554,44 @@ nfs_validate_shareopts(const char *shareopts)
...
-			fclose(nfs_exportfs_temp_fp);
-			return (B_TRUE);
+	while ((getline(&buf, &buflen, fp)) != -1) {
+		char *space = NULL;
+
+		if ((space = strchr(buf, ' ')) != NULL) {
+			if (strncmp(impl_share->sa_mountpoint, buf,
+			    space - buf) == 0 &&
+			    strncmp(impl_share->sa_mountpoint, buf,
+			    strlen(impl_share->sa_mountpoint)) == 0) {
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:580]: Same suggestion as in nfs_copy_entries().

File: lib/libshare/os/linux/nfs.c [hunk 19]
```diff
@@ -512,118 +554,44 @@ nfs_validate_shareopts(const char *shareopts)
...
+				fclose(fp);
+				return (B_TRUE);
```
Comment (ahrens) [inline lib/libshare/os/linux/nfs.c:582]: don't we need to `free(buf)` here?

File: lib/libzfs/libzfs_changelist.c [hunk 20]
```diff
@@ -167,7 +171,7 @@ changelist_postfix(prop_changelist_t *clp)
 	uu_avl_walk_t *walk;
 	char shareopts[ZFS_MAXPROPLEN];
 	int errors = 0;
-	libzfs_handle_t *hdl;
+	boolean_t update_shares = B_FALSE;
```
Comment (ahrens) [inline lib/libzfs/libzfs_changelist.c:174]:
```suggestion
	boolean_t commit_shares = B_FALSE;
```

File: lib/libshare/os/linux/nfs.c (hunk 4 above)
Comment (grwilson) [reply lib/libshare/os/linux/nfs.c:56]: Yes, the `sharenfs` property still uses the Solaris option in linux and converts them to something that can be used by exportfs. I don't think all Solaris options are supported.

**Comment by ahrens** [comment]
From talking to @grwilson, this PR also fixes #7943.

File: lib/libshare/os/linux/nfs.c [hunk 21]
```diff
@@ -634,114 +626,45 @@ nfs_update_shareopts(sa_share_impl_t impl_share, const char *resource,
...
-	free(FSINFO(impl_share, nfs_fstype)->shareopts);
 	FSINFO(impl_share, nfs_fstype)->shareopts = NULL;
 }
 
+static int
+nfs_commit_shares(void)
+{
+	char *argv[] = {
+	    "/usr/sbin/exportfs",
+	    "-r",
```
[the remark this answers is not in the document]
Comment (grwilson) [reply lib/libshare/os/linux/nfs.c:637]: I was more interested in the use case. I'm assuming that the concern is that there might be filesystems which are shared without being persisted in `/etc/exports` and as a result they would not get reshared. Is that the use case?

File: lib/libshare/os/freebsd/nfs.c [hunk 22]
```diff
@@ -0,0 +1,394 @@
...
+	 */
+	FILE *oldfp = fopen(ZFS_EXPORTS_FILE, "r");
+	if (oldfp == NULL)
+		return (SA_OK);
+
+	FILE *newfp = fopen(filename, "w+");
+	fprintf(newfp, "%s", FILE_HEADER);
+	while (oldfp != NULL && (line = zgetline(oldfp, mountpoint)) != NULL)
+		fprintf(newfp, "%s\n", line);
+	if (oldfp != NULL && ferror(oldfp) != 0) {
```
[the remark this answers is not in the document]
Comment (grwilson) [reply lib/libshare/os/freebsd/nfs.c:226]: I copied this from libzfs_fsshare but I don't believe that it can happen. I'll remove and retest.

File: lib/libshare/os/linux/nfs.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (bgly) [reply lib/libshare/os/linux/nfs.c:637]: The use case for me was committing shares. For example you add a bunch of shares and you want to notify NFS service to go re-read and update the tables with the new shares or deleted shares you run the exportfs -ra command. If you notice the FREEBSD implementation you have the same function called: nfs_commit_shares that goes and restarts mountd service. That does the same thing as exportfs -ra. My main concern is that so far without your patch sets, right now when you run sharenfs a bunch of times not in parallel some shares get stale file handles on client side. I didnt look closely at the prior implementation, but I was suggesting to add the `a` to make sure everything is in sync between the file nfs server uses to the kernel export table. I also briefly tested your patch set with: 

1. Create 10 volumes with default shares, so not shared to any external ip's
2. Update the shares with a couple of ip's
3. Manually mount just 1 shares, do a DD
4. While DD is running update all the shares in parallel, Also make sure remove the one ip that is not connected, but also associated to the same share doing the DD
5. See if you get stale file handle, in this case it did not BUT not all the shares got updated.
6. check `showmount -e`
So in conclusion, when i updated 10 shares in parallel, one of them removing an ip, the rest adding an ip, only 3 out of the 10 shares got updated.

File: lib/libshare/os/freebsd/nfs.c [hunk 23]
```diff
@@ -0,0 +1,436 @@
...
+ *	-ro -network=192.168.0.0 -mask=255.255.255.0 -maproot=0 freefall.freebsd.org
+ *
+ * Recognized keywords:
+ *
+ *	ro, maproot, mapall, mask, network, sec, alldirs, public, webnfs, index, quiet
+ *
+ */
+/* END CSTYLED */
+
+static const char *known_opts[] = { "ro", "maproot", "mapall", "mask",
```
Comment (ahrens) [inline lib/libshare/os/freebsd/nfs.c:148]: Could `known_opts` be defined inside `translate_opts()`?  Not sure if that is a good idea or not, style-wise, but it seems like it's only used by that func.  And we are already using static locals inside the func (`newopts`).

File: lib/libshare/os/freebsd/nfs.c [hunk 24]
```diff
@@ -0,0 +1,436 @@
...
+	"network", "sec", "alldirs", "public", "webnfs", "index", "quiet",
+	NULL };
+static char *
+translate_opts(const char *shareopts)
```
Comment (ahrens) [inline lib/libshare/os/freebsd/nfs.c:152]: Maybe we should add a warning that this is not thread-safe, becuase it returns a `static` buffer.

File: lib/libshare/os/freebsd/nfs.c [hunk 25]
```diff
@@ -0,0 +1,436 @@
...
+		unlink(tmpfile);
+		free(tmpfile);
+		return (SA_SYSTEM_ERR);
+	}
+	free(tmpfile);
+	return (SA_OK);
+}
+
+static int
+nfs_copy_entries(char *filename, const char *mountpoint)
```
Comment (ahrens) [inline lib/libshare/os/freebsd/nfs.c:217]: add a comment explaining the semantics?  `Copy entries from the exports file to "filename", omitting any entries for the specified mountpoint.`

(these suggestions may apply to `linux/nfs.c` as well)

File: tests/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_001_pos.ksh [hunk 26]
```diff
@@ -138,11 +140,21 @@ done
 #
 log_must zfs share -a
 
+#
+# We need to unset __ZFS_POOL_EXCLUDE so that we include all file systems
+# into the /etc/exports.d/zfs.exports file. This will be get reset
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_001_pos.ksh:145]: "into the" -> "in the"
"This will be get reset" -> "This will be reset"

File: tests/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_concurrent_shares.ksh [hunk 27]
```diff
@@ -0,0 +1,203 @@
...
+
+#
+# Main test routine.
+#
+# Given a file system this routine will attempt
+# share the mountpoint and then verify it has been shared.
+#
+function test_share # filesystem
+{
+	set -x
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_concurrent_shares.ksh:98]: You'll probably want to remove this in the final version.

File: tests/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_concurrent_shares.ksh [hunk 28]
```diff
@@ -0,0 +1,203 @@
...
+#
+# Try a zfs share -a and verify all file systems are shared.
+#
+log_must zfs share -a
+
+#
+# We need to unset __ZFS_POOL_EXCLUDE so that we include all file systems
+# into the /etc/exports.d/zfs.exports file. This will get reset by
+# the next test.
+#
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_concurrent_shares.ksh:183]: "into the" -> "in the"
"This will be get reset" -> "This will be reset"

**Comment by freqlabs** [comment]
@grwilson The failure in zfs_share_concurrent_shares on stable/12 seems to have been pretty consistent. Do you have any indication of what this is about? I was hoping it would be resolved by https://reviews.freebsd.org/rS362713 but that should be in the image used by the bot as of the 2nd.

**Comment by grwilson** [comment]
> @grwilson The failure in zfs_share_concurrent_shares on stable/12 seems to have been pretty consistent. Do you have any indication of what this is about? I was hoping it would be resolved by https://reviews.freebsd.org/rS362713 but that is in the image used by the bot as of the 2nd.

I've have reproduced this on freebsd 12.1/stable (build r363017) and noticed that the output of `showmount -E` (which is used by the test suite to detect if the filesystem is shared) does not show the accurate count immediately. I ran a simple test where I shared 1000 filesystems in parallel and immediately ran `showmount -E` and it displayed 994 filesystem but rerunning it again displays the correct value. I suspect that `mountd` is being delayed because of the number of updates happening during the test and so we don't get an inaccurate result from `showmount -E` and fail the test.

I have marked this test as flakey in freebsd since it does work on 12.1/release.
